lib/pbench/server/database/models/datasets.py

```diff
@@ -944,24 +971,38 @@ def setvalue(dataset: Dataset, key: str, value: Any) -> "Metadata":
             properly assigned value.
         """
         keys = key.lower().split(".")
+        if "" in keys:
+            raise MetadataBadKey(key)
         native_key = keys.pop(0)
         found = True
         try:
             meta = Metadata.get(dataset, native_key)
-            meta_value = meta.value.copy()
+
+            # SQLAlchemy determines whether to perform an `update` based on the
+            # Python object reference. We make a copy here to ensure that it
+            # sees we've made a change.
+            meta_value = copy.deepcopy(meta.value)
         except MetadataNotFound:
             found = False
             meta_value = {}
         if not keys:
             meta_value = value
         else:
+            last_key = keys[-1]
             hierarchy = meta_value
```

Comment:
But `meta_value` descends into the structure while `meta_value` remains at the top, because that's what we need to reflect into the DB. I strained my brain for at least a tenth of a second hoping for a better name to pop out, but it didn't.

Reply:
(I assume that, in your reply above, your first reference to `meta_value` should actually be `hierarchy`....)

I suppose that you could leave things as they are and then set `meta_value` equal to `hierarchy` after the decent.  That would be a no-op, but it would make the code clear.

Reply:
> But `hierarchy` descends into the structure while `meta_value` remains at the top

Quite right.  I was confused.  My apologies!  Nevermind...although `walker` is a nice improvement over `hierarchy`, so that's a win.